solr/core/src/java/org/apache/solr/response/transform/DeeplyNestedChildDocTransformer.java

```diff
@@ -0,0 +1,214 @@
+    this.name = name;
+    this.schema = req.getSchema();
+    this.idField = this.schema.getUniqueKeyField();
+    this.req = req;
+    this.parentsFilter = parentsFilter;
+    this.limit = limit;
+    this.childFilterQuery = childFilterQuery;
+  }
+
+  @Override
+  public String getName()  {
+    return name;
+  }
+
+  @Override
+  public String[] getExtraRequestFields() {
+    // we always need the idField (of the parent) in order to fill out it's children
+    return new String[] { idField.getName() };
+  }
+
+  protected static String getSolrFieldString(Object fieldVal, FieldType fieldType) {
+    return fieldVal instanceof IndexableField
+        ? fieldType.toExternal((IndexableField)fieldVal)
+        : fieldVal.toString();
+  }
+
+  @Override
+  public void transform(SolrDocument rootDoc, int rootDocId) {
+
+    FieldType idFt = idField.getType();
+
+    String rootIdExt = getSolrFieldString(rootDoc.getFirstValue(idField.getName()), idFt);
+
+    try {
+      Query parentQuery = idFt.getFieldQuery(null, idField, rootIdExt);
+      Query query = new ToChildBlockJoinQuery(parentQuery, parentsFilter);
+      SolrIndexSearcher searcher = context.getSearcher();
+      DocList children = searcher.getDocList(query, childFilterQuery, docKeySort, 0, limit);
+      long segAndId = searcher.lookupId(new BytesRef(rootIdExt));
+      final int seg = (int) (segAndId >> 32);
+      final LeafReaderContext leafReaderContext = searcher.getIndexReader().leaves().get(seg);
+      final SortedDocValues segPathDocValues = DocValues.getSorted(leafReaderContext.reader(), NEST_PATH_FIELD_NAME);
+
+      Multimap<String,SolrDocument> pendingParentPathsToChildren = ArrayListMultimap.create();
+
+      if(children.matches() > 0) {
+        SolrDocumentFetcher docFetcher = searcher.getDocFetcher();
+        Set<String> dvFieldsToReturn = docFetcher.getNonStoredDVs(true).stream()
+            .filter(name -> !NEST_PATH_FIELD_NAME.equals(name)).collect(Collectors.toSet());
+        boolean shouldDecorateWithDVs = dvFieldsToReturn.size() > 0;
+        DocIterator i = children.iterator();
+        final int segBaseId = leafReaderContext.docBase;
+        final int firstChildDocId = i.nextDoc();
+        assert firstChildDocId < rootDocId;
+
+        for (int docId = firstChildDocId; docId < rootDocId; ++docId) {
+          // get the path
+          String fullDocPath = getPathByDocId(docId - segBaseId, segPathDocValues);
+
+          // Is this doc a direct ancestor of another doc we've seen?
+          boolean isAncestor = pendingParentPathsToChildren.containsKey(fullDocPath);
+
+          // Do we need to do anything with this doc (either ancestor or a matched the child query)
+          if (isAncestor || children.exists(docId)) {
+            // load the doc
+            SolrDocument doc = DocsStreamer.convertLuceneDocToSolrDoc(docFetcher.doc(docId),
+                schema, new SolrReturnFields());
+            doc.setField(NEST_PATH_FIELD_NAME, fullDocPath);
```

Comment:
This approach kinda requires we remove such internal fields later.

Reply:
It's probably easiest to remove it later, probably at the point we add children to document.  Or try the Multimap<String<MultiMap<String,SolrDocument>> idea I had, wherein the intermediate String is the label.

Reply:
Perhaps we can Store everything in a Multimap<String,Pair<String,SolrDocument>>, thus saving each document bonded to its full path?

Reply:
A Pair would be fine too; it's only slightly more bulky than an intermediate Multimap.

Reply:
It seems like a MultiMap<String, Pair<String,Pair>> is even more efficient, since each path is unique, while the inner MultiMap will store each item in each child label as another array.

Reply:
I expected the intermediate string is a *label* like “comment” not a path. A label will be shared by all children under that label.

Reply:
Oh, ok.  Perhaps have a separate standalone Set of complete paths of "single child" elements.  that tracks parent path to child labels in which that label contains a unique entry.  WDYT?

Reply:
To be clear, I just mean a `Set<String>` of paths like `foo#1/bar#9/author#`  (author is single valued)